### PR TITLE
415 populate fields in edit component window

### LINKED ISSUE #415: Populate fields in edit component window
Follows on from #414 

The edit component window loads everything but the component's fields currently. I did not put this in #414 because it is not a requirement for #300.

### LINKED ISSUE #497: Remove buffer size options from ev42 stream options and add source
The following options are no longer implemented in the file writer `ev42` writer module:
- `nexus.buffer.size_mb`
- `nexus.buffer.size_kb`
- `nexus.buffer.packet_max_kb`

### PR DESCRIPTION
**Pull request by rerpha**
### Issue

Closes #415 
Closes #497 

### Description of work
Adds a way to edit the fields in the edit component window. Previously these were not constructed when editing components. 

~If this is merged before #423 we will need to add in the ability to edit stream widgets too.~ - done

### Acceptance Criteria 

Should be able to edit scalar, array, link and stream fields now when editing components. 

This should also work on file load. 

### UI tests

Added 

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

**Comment by rerpha** [comment]
TODO: 
sort merge conflict out 
add streams editing

**Comment by rerpha** [comment]
still needs to do UI tests, not ready for review yet

File: nexus_constructor/nexus/nexus_wrapper.py [hunk 1]
```diff
@@ -187,6 +187,24 @@ def get_field_value(group: h5py.Group, name: str) -> Optional[Any]:
...
+                fields.append(item)
+            if isinstance(item, h5py.Group):
+                if isinstance(item.parent.get(item.name, getlink=True), h5py.SoftLink):
+                    fields.append(item)
+                if (
+                    "NX_class" in item.attrs.keys()
+                    and item.attrs["NX_class"] == "NCstream"
+                ):
+                    fields.append(item)
+        return fields
```
Comment (DominicOram) [inline nexus_constructor/nexus/nexus_wrapper.py]: The filtering this function does is basically the same as the filtering done in `add_component_window` can we avoid doing it twice?

File: nexus_constructor/stream_advanced_options.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+NEXUS_INDICES_INDEX_EVERY_MB = "nexus.indices.index_every_mb"
+NEXUS_INDICES_INDEX_EVERY_KB = "nexus.indices.index_every_kb"
+STORE_LATEST_INTO = "store_latest_into"
+NEXUS_CHUNK_CHUNK_MB = "nexus.chunk.chunk_mb"
+NEXUS_CHUNK_CHUNK_KB = "nexus.chunk.chunk_kb"
+NEXUS_BUFFER_SIZE_MB = "nexus.buffer.size_mb"
+NEXUS_BUFFER_SIZE_KB = "nexus.buffer.size_kb"
+NEXUS_BUFFER_PACKET_MAX_KB = "nexus.buffer.packet_max_kb"
+ADC_PULSE_DEBUG = "adc_pulse_debug"
```
Comment (DominicOram) [inline nexus_constructor/stream_advanced_options.py]: 👍

File: tests/ui_tests/test_ui_add_component_window.py [hunk 3]
```diff
@@ -1338,6 +1338,172 @@ def test_UI_GIVEN_component_with_cylinder_shape_WHEN_editing_component_THEN_cyli
             assert dialog.cylinderOptionsBox.isEnabled()
 
 
+def test_UI_GIVEN_component_with_scalar_field_WHEN_editing_component_THEN_field_appears_in_fields_list_with_correct_value(
+    qtbot
+):
+    instrument = Instrument(NexusWrapper("test_component_editing_scalar_field"))
+    model = ComponentTreeModel(instrument)
+    component_name = "chopper1"
+    component = instrument.create_component(component_name, "NXdisk_chopper", "")
```
Comment (DominicOram) [inline tests/ui_tests/test_ui_add_component_window.py]: These lines are basically the same in all the tests, can we just put into a helper method?

File: nexus_constructor/add_component_window.py [hunk 4]
```diff
@@ -251,9 +265,147 @@ def _fill_existing_entries(self):
...
+                field[NEXUS_INDICES_INDEX_EVERY_MB][()]
+            )
+        if NEXUS_INDICES_INDEX_EVERY_KB in field.keys():
+            new_ui_field.streams_widget.f142_nexus_indices_index_every_kb_spinbox.setValue(
+                field[NEXUS_INDICES_INDEX_EVERY_KB][()]
+            )
+        if STORE_LATEST_INTO in field.keys():
+            new_ui_field.streams_widget.f142_nexus_store_latest_into_spinbox.setValue(
+                field[STORE_LATEST_INTO][()]
+            )
```
[the remark this answers is not in the document]
Comment (rerpha) [reply nexus_constructor/add_component_window.py]: have moved them into free functions instead so can be put anywhere

**Comment by rerpha** [comment]
todo: clear all fields that aren't in the field blacklist before adding edited fields back into the group

File: nexus_constructor/nexus/nexus_wrapper.py [hunk 5]
```diff
@@ -43,6 +43,34 @@ def decode_bytes_string(nexus_string):
...
+    """
+    scalar_fields = []
+    array_fields = []
+    stream_fields = []
+    link_fields = []
+    for item in group.values():
+        if isinstance(item, h5py.Dataset) and item.name.split("/")[-1] != "description":
+            if np.isscalar(item.value):
+                scalar_fields.append(item)
+                break
```
Comment (DominicOram) [inline nexus_constructor/nexus/nexus_wrapper.py]: Do you mean `continue` here? I don't think you want to completely stop the loop?

File: nexus_constructor/add_component_window.py [hunk 6]
```diff
@@ -299,29 +313,17 @@ def _fill_existing_entries(self):
...
-
-        for field in stream_fields:
-            new_ui_field = self.create_new_ui_field(field)
-            new_ui_field.field_type = FieldType.kafka_stream.value
-            new_ui_field.streams_widget.update_existing_stream_info(field)
-
-        for field in link_fields:
-            new_ui_field = self.create_new_ui_field(field)
-            update_existing_link_field(field, new_ui_field)
+        update_methods = {
```
Comment (rerpha) [inline nexus_constructor/add_component_window.py]: cool :+1:

**Comment by rerpha** [comment]
Will need to remove the stream name stuff when #491 is closed

**Pull request merged**